app/src/main/java/io/github/dovecoteescapee/byedpi/fragments/VpnAppsFilterFragment.kt

```diff
@@ -0,0 +1,131 @@
+    private lateinit var checked: MutableSet<String>
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+        binding = FragmentFilterBinding.inflate(inflater, container, false)
+
+        val activity = requireActivity()
+
+        checked = activity.getPreferences().getStringSet("vpn_filtered_apps", emptySet())!!
+            .toMutableSet()
+        adapter = AppItemAdapter(
+            activity,
+            { checked.contains(it.packageName) },
+            { app, status ->
+                checked.apply {
+                    if (status) {
+                        add(app.packageName)
+                    } else {
+                        remove(app.packageName)
+                    }
+                }
+                activity.getPreferences().edit {
+                    putStringSet("vpn_filtered_apps", checked.toSet())
+                }
+            })
+
+
+        binding.appList.layoutManager = LinearLayoutManager(activity)
+        binding.appList.addItemDecoration(
+            DividerItemDecoration(
+                binding.appList.context,
+                RecyclerView.VERTICAL
+            )
+        )
+
+        binding.appList.adapter = adapter
+
+        Thread {
+            apps = collectApps()
+
+            activity.runOnUiThread {
+                adapter.setList(apps)
+
+                binding.progressCircular.visibility = View.GONE
+                binding.appList.visibility = View.VISIBLE
+            }
+        }.start()
+
+        requireActivity().addMenuProvider(this, viewLifecycleOwner, Lifecycle.State.RESUMED)
+
+        return binding.root
+    }
+
+    private fun collectApps(): List<AppItemAdapter.Model> {
+        val context = requireContext()
+        val packageManager = context.packageManager
+
+        return packageManager.getInstalledApplications(0)
+            .filter { it.packageName != context.packageName }
+            .map {
+                AppItemAdapter.Model(
+                    label = packageManager.getApplicationLabel(it).toString(),
+                    packageName = it.packageName,
+                    icon = packageManager.getApplicationIcon(it),
+                )
+            }
+    }
+
+    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
+        menuInflater.inflate(R.menu.menu_filter, menu)
+
+        val searchItem = menu.findItem(R.id.action_search)
+        (searchItem.actionView as SearchView).setOnQueryTextListener(object : OnQueryTextListener {
+            override fun onQueryTextSubmit(query: String): Boolean {
+                val lQuery = query.lowercase()
+                adapter.setList(apps.filter { it.label.lowercase().contains(lQuery) })
+                return false
+            }
+
+            override fun onQueryTextChange(query: String?): Boolean {
+                return false
```

Comment:
Why not update on text change? I think it would be more convenient and shouldn't work well

Reply:
In this case, debouncing is needed to avoid performance problems

Reply:
For some reason I didn’t notice that there was a dependency оn lifecycle package for coroutines, let's also save the job context so that it can be canceled on the next query update

Reply:
> In this case, debouncing is needed to avoid performance problems

I don't think it's a problem for modern devices. There are usually not many installed applications. Many VPN apps update list on edit, for example, AmneziaVPN, Psiphon, Lantern, Orbot

> let's also save the job context so that it can be canceled on the next query update

Yes, it's good idea